src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+        /// <param name="payload">A JSON representation of an HttpRequest payload.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>SignedHttpRequest as a JWS in Compact Serialization Format.</returns>
+        protected virtual Task<string> SignHttpRequestAsync(string header, string payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(header))
+                throw LogHelper.LogArgumentNullException(nameof(header));
+
+            if (string.IsNullOrEmpty(payload))
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var message = $"{Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header))}.{Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(payload))}";
+            var signature = JwtTokenUtilities.CreateEncodedSignature(message, signedHttpRequestDescriptor.SigningCredentials);
+            return Task.FromResult($"{message}.{signature}");
+        }
+
+        /// <summary>
+        /// Adds the 'at' claim to the <paramref name="payload"/>.
+        /// </summary>
+        /// <param name="payload">HttpRequest payload represented as a <see cref="Dictionary{TKey, TValue}"/>.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        protected virtual void AddAtClaim(Dictionary<string, object> payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            payload.Add(SignedHttpRequestClaimTypes.At, signedHttpRequestDescriptor.AccessToken);
+        }
+
+        /// <summary>
+        /// Adds the 'ts' claim to the <paramref name="payload"/>.
+        /// </summary>
+        /// <param name="payload">HttpRequest payload represented as a <see cref="Dictionary{TKey, TValue}"/>.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <remarks>
+        /// This method will be executed only if <see cref="SignedHttpRequestCreationParameters.CreateTs"/> is set to <c>true</c>.
+        /// </remarks>    
+        protected virtual void AddTsClaim(Dictionary<string, object> payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var signedHttpRequestCreationTime = DateTime.UtcNow.Add(signedHttpRequestDescriptor.SignedHttpRequestCreationParameters.TimeAdjustment);
```

Comment:
Can we use EpochTime.cs?

Reply:
TimeAdjustment is a TimeSpan. I'm not sure how to use it with EpochTime.
I'm using EpochTime to calculate the 'ts' claim.

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/541bfca2bbe83d65db4df44f783b2c966f1fd67b/src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs#L196-L197

Are you suggesting that I should replace ln. 197 with:

```csharp
payload.Add(SignedHttpRequestClaimTypes.Ts, EpochTime.GetIntDate(signedHttpRequestCreationTime));
```

Reply:
todo: the assumption is correct.